File: reHw/1/list.cpp
#include "list.h"
#include <iostream>
using namespace std;

List :: add( int x)
{
    ListElement * tmp = new ListElement (x, NULL);
    tmp->setNext(head->getNext());
    head->setNext(tmp);
}

List :: ~List()
{
    while (head->getNext != NULL)
    {
        ListElement *tmp = head;
        head->setNext(head->getNext());
        delete tmp;
    }
    delete head;
}

List ::valueHead()
{
    return head->setValue();
}

List :: print()
{
    ListElement * tmp = head;
    while (tmp->getNext() != NULL)
    {
        tmp->setNext(tmp->getNext());
        cout << tmp->setValue();
    }
}

File: C#/7/hw7/hw7t1/List.cs
using System;
using System.Collections.Generic;

namespace hw7t1
{
    public class List<T>: IEnumerable<T>
    {

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return new Iterator<T>(this);
        }

        IEnumerator<T> IEnumerable<T>.GetEnumerator()
        {
            return new Iterator<T>(this);
        }

        public class ListElement
        {
            public ListElement()
            { }

            public ListElement (T v)
            {
                next = null;
                value = v;
            }

            public ListElement GetNext()
            {
                return next;
            }

            public void SetNext(ListElement nxt)
            {
                next = nxt;
            }

            public T GetValue()
            {
                return value;
            }

            private T value;
            private ListElement next;
        }

        public List()
        {
            head = new ListElement();
        }

        public void Add(T val)
        {
            ListElement l = head;
            while (l.GetNext() != null)
            {
                l = l.GetNext();
            }
            ListElement tmp = new ListElement(val);
            tmp.SetNext(l.GetNext());
            l.SetNext(tmp);
        }

        public T Del()
        {
            if (head.GetNext() == null)
                throw new ListException();
            ListElement tmp = head.GetNext();
            head.SetNext(tmp.GetNext());
            return tmp.GetValue();
        }
        
        public ListElement head;
    }
}
